Use sudo enabled Trusty for HHVM

photodude opened the pull request:
This provides the current HHVM version (3.15.2 as of this PR) and will track with each release (i.e. will be 3.16 when 3.16 is released.

If testing against HHVM LST versions is desired follow this guide. https://docs.travis-ci.com/user/languages/php#HHVM-versions

Should be able to change to container based Trusty after Q1-17 https://blog.travis-ci.com/2016-11-08-trusty-container-public-beta/

photodude committed:
Use sudo enabled Trusty for HHVM

nazar-pc commented on .travis.yml:
```diff
@@ -17,12 +16,33 @@ env:
+            - mysql-server-5.6
+            - mysql-client-core-5.6
+            - mysql-client-5.6
+      services:
```
Why do we need empty list?

nazar-pc commented on .travis.yml:
```diff
@@ -34,9 +54,10 @@ before_script:
     fi
   - |
-    if [[ $TRAVIS_PHP_VERSION != 'hhvm' ]]; then
+    if [[ $TRAVIS_PHP_VERSION != 'hhv*' ]]; then
```
Don't think this will work, why even changing it (plain hhvm is used in above config)?

photodude replied:
This is the format I use in my repos and those that I contribute to. By using `hhv*`  it covers `hhvm`, `hhvm-nightly` and LTS specified versions like `hhvm-3.12`

Although I did forget to remove the tick marks which does cause a problem.

nazar-pc reviewed (commented):
Generally I'm fine with testing on HHVM, but there are 2 caveats here:
- until Phar write support implemented in HHVM not all code will be tested
- `sudo: true` builds take significant amount of time (I know there is `fast_finish: true` now) while not providing too much value

What is your interest in getting HHVM here?

photodude commented:
> What is your interest in getting HHVM here?

No actual interest in the project. I just saw an issue you had filed with Travis CI and noticed it was testing against the precise hhvm which is crazy out of date. Where I can I help projects testing against the old version start testing against the current version. For me, this improves testing against HHVM on travis and hopefully increases HHVM acceptance rates.

> sudo: true builds take significant amount of time

I do agree, that "sudo: true builds take 'a longer' amount of time" compared to the container testing. Current HHVM is only available on Trusty and the Beta container is not yet working well enough to transition to that (Trusty container release projected in Q1-17). It's important to note that the only part being tested with `sudo: true` in this scenario is HHVM, everything else is still using `sudo: false`

photodude committed:
memcached is preinstalled on HHVM

- We don't need to add memcached to hhvm ini config as it's preinstalled.
- fix tick marks on conditionals for `hhv*`
- remove empty services lines from hhvm

photodude committed:
use mysql root (required for trusty)

must use  mysql root for testing on trusty, does not affect other build options

photodude committed:
travis database exists on hhvm

looks like we don't need to create the travis database for hhvm on trusty sudo required

nazar-pc commented on .travis.yml:
```diff
@@ -17,14 +16,36 @@ env:
+
 before_script:
-  - mysql -e 'CREATE DATABASE `travis`;'
+  - |
```
Precise image uses `travis` user, why `root` user is needed here?

photodude replied:
It's easier (simpler) to just run everything with root rather than adding additional lines to only use root for trusty.

photodude replied:
You can read the reason for root in the travis docs
https://docs.travis-ci.com/user/database-setup/#MySQL-5.6

nazar-pc replied:
Well, for compatibility with tests `travis` user should be created, since it is hardcoded in tests.
And this command should be different for Precise and Trusty.

nazar-pc replied:
If there is access to root user and possibility to create database - it should be possible to create user as well.

photodude replied:
I created the user, but the related failures were not solved. The following message is in the tests
```php
Access denied for user 'travis'@'localhost' to database 'travis'
```

I tried granting permissions but that also didn't solve the errors, I just got new messages
```php
Warning: Argument 1 not passed as reference in /core/drivers/DB/MySQLi.php on line 118
Fatal error: Call to undefined method mysqli_stmt::get_result() in /core/drivers/DB/MySQLi.php on line 123
```

nazar-pc replied:
Tweaked few more things and HHVM with SQLite passes: https://travis-ci.org/nazar-pc/CleverStyle-Framework/jobs/175160579
Thank you very much for support!

photodude replied:
You are welcome, it's good to see one more project testing against the current version of hhvm.

nazar-pc commented on .travis.yml:
```diff
@@ -33,10 +54,7 @@ before_script:
-    if [[ $TRAVIS_PHP_VERSION != 'hhvm' ]]; then
-      echo "extension = memcached.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
-    fi
+
```
This is not about HHVM, PHP does need to have memcached inserted into config.

photodude replied:
This has been restored.

nazar-pc reviewed (commented):
Few notes since there is a mix of Precise/Trusty images now.
Also nice trick with `hhv*`, didn't know it:)

nazar-pc commented:
To be honest, Travis CI is stagnating, they should be moving to 16.04 instead of 2.5 years old 14.04 which has already outdated software.

photodude commented:
I somewhat agree. Unfortunately they are not even starting work on 16.04 until Q2-17 since according to travis CI 14.04 was "a rather significant undertaking, going beyond offering Trusty images on EC2"

Hopefully, that means 16.04 will be simpler and faster for them in implement after work starts.

Regardless of whether "14.04 has already outdated software." it still provides a useful testing platform for testing against things like current hhvm.

photodude committed:
restore memcached for non-hhvm tests

photodude committed:
try creating a travis user for trusty hhvm

nazar-pc committed:
Simplify condition
